crates/walrus-service/src/node/consistency_check.rs

```diff
@@ -72,26 +117,96 @@ fn get_epoch_bucket(epoch: Epoch) -> String {
-    >,
+fn compose_blob_list_digest_and_check_sliver_data_existence(
+    node: &StorageNodeInner,
+    blob_info_iter: BlobInfoIterator<'_>,
     epoch: Epoch,
     scan_counter: &IntCounterVec,
-) -> Result<u64, TypedStoreError> {
+    blobs_not_yet_fully_synced: &HashSet<BlobId>,
+) -> Result<BlobConsistencyCheckResult, TypedStoreError> {
+    // Create a new tokio runtime for the async task to check blob existence.
+    #[cfg(not(msim))]
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .build()
+        .expect("failed to create tokio runtime");
+
     let epoch_bucket = get_epoch_bucket(epoch);
+    let enable_sliver_data_existence_check = node
+        .consistency_check_config
+        .enable_sliver_data_existence_check;
 
     // xxhash is not a cryptographic hash function, but it is fast, has good collision
     // resistance, and is consistent across all platforms.
     let mut hasher = twox_hash::XxHash64::with_seed(epoch as u64);
+    let mut total_synced_scanned = 0;
+    let mut total_fully_stored = 0;
+    let mut existence_check_error = 0;
     for item in blob_info_iter {
         match item {
             Ok(blob_info) => {
                 hasher.write(blob_info.0.as_ref());
+
+                if enable_sliver_data_existence_check
+                    && !blobs_not_yet_fully_synced.contains(&blob_info.0)
+                {
+                    total_synced_scanned += 1;
+
+                    #[cfg(not(msim))]
+                    {
+                        handle_existence_check_result(
+                            // Note that we are running the async function in a new reusable Runtime
+                            // created at the beginning of the task. Note that performance of this
+                            // task is not a concern, and we should limit the resource this task
+                            // can use.
+                            //
+                            // TODO(zhewu): ideally, we should create iterators over the sliver
+                            // column families, and perform sequential scan along with
+                            // BlobInfoIterator to conduct more efficient existence check. This
+                            // requires the SafeIterator to support seek() functionality first.
+                            rt.block_on(node.is_stored_at_all_shards_at_latest_epoch(&blob_info.0)),
+                            &mut total_fully_stored,
+                            &mut existence_check_error,
+                            &blob_info.0,
+                        );
+                    }
+
+                    // Unfortunately, msim does not support tokio::Runtime::block_on, so we need to
+                    // use less ideal method to execute the async function that checks blob
+                    // data existence. Note that simtest is mostly for correctness verification, so
+                    // the performance is not a concern.
```

Comment:
I spent on this issue quite a long time, and this is the best solution I can think of at the moment.